### PR TITLE
[Spec Decode] DSpark speculators checkpoint support

### PR DESCRIPTION
**Pull request by mgoin**
Validated with [Qwen3-8B](https://huggingface.co/mgoin/Qwen3-8B-speculator.dspark-reasoning) and [GLM 5.2 DSpark (preview)](https://huggingface.co/mgoin/GLM-5.2-speculator.dspark-preview#greedy-temperature0)

```
vllm serve zai-org/GLM-5.2-FP8 -tp 8 \
  --speculative-config '{"method":"dspark","model":"mgoin/GLM-5.2-speculator.dspark-preview","num_speculative_tokens":7,"attention_backend":"FLASH_ATTN","draft_sample_method":"probabilistic"}'
```

### CODE REVIEW

File: vllm/model_executor/models/qwen3_dspark.py [hunk 1]
```diff
@@ -36,18 +36,26 @@
...
+        self,
+        vocab_size: int,
+        draft_vocab_size: int,
+        markov_rank: int,
+        prefix: str,
+    ) -> None:
         super().__init__()
         # TODO(ben): profile for which (if any) it makes sense to replicate or TP-shard
         self.markov_w1 = VocabParallelEmbedding(
             vocab_size, markov_rank, prefix=maybe_prefix(prefix, "markov_w1")
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dspark.py:55]: Why does this project from the vocab_size and not draft_vocab_size?

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dspark.py:55]: Would it be equivalent to just run the markov entirely on the vocab subspace?

Comment (mgoin) [reply vllm/model_executor/models/qwen3_dspark.py:55]: I think this is because markov_w1 embeds the conditioning token (aka the previous token). Here are the checkpoint shapes
```
markov_head.markov_w1.weight   (151936, 256)
markov_head.markov_w2.weight   (32000, 256)
lm_head.weight                 (32000, 4096)
```
Since the conditioning token can be any token, it needs the full vocab. Only markov_w2 produces the bias for the draft logits.

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dspark.py:55]: ah, right

File: vllm/model_executor/models/qwen3_dspark.py [hunk 2]
```diff
@@ -117,6 +129,17 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = "") -> None:
     def get_draft_kv_cache_layer_names(self) -> list[str]:
         return [layer.self_attn.attn.layer_name for layer in self.model.layers]
 
+    def compute_draft_logits(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        # Draft-vocab logits without the d2t scatter: the speculator adds the
+        # Markov bias in draft space, then remaps via map_draft_to_target.
+        return self.logits_processor(self.lm_head, hidden_states)
+
+    def map_draft_to_target(self, draft_ids: torch.Tensor) -> torch.Tensor:
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dspark.py:137]: This has 1 call site AFAICT, does it need to be a helper? I don't think any of the other spec model impls use a helper like this?

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dspark.py:137]: I guess it's a contract between the model and the speculator implementation. Since it's usually handled in compute_logits. Makes sense I suppose. Would be nice to unify this interface, though it's fine for now

File: vllm/transformers_utils/configs/speculators/algos.py [hunk 3]
```diff
@@ -120,3 +120,46 @@ def update_dflash(config_dict: dict, pre_trained_config: dict) -> None:
...
+    pre_trained_config["draft_vocab_size"] = config_dict.get("draft_vocab_size")
+    if config_dict.get("target_hidden_size") is not None:
+        pre_trained_config["target_hidden_size"] = config_dict["target_hidden_size"]
+
+    aux_layer_ids = config_dict["aux_hidden_state_layer_ids"]
+    pre_trained_config["eagle_aux_hidden_state_layer_ids"] = aux_layer_ids
+    # DSpark indexes target layers as aux_id - 1 (matches the dense configs).
+    pre_trained_config["target_layer_ids"] = [i - 1 for i in aux_layer_ids]
+
+    for key in (
```
Comment (benchislett) [inline vllm/transformers_utils/configs/speculators/algos.py:155]: nit: I think `target_hidden_size` can go in this list? and maybe draft_vocab_size also?

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 4]
```diff
@@ -38,8 +40,15 @@ class DSparkSpeculator(DFlashSpeculator):
     def __init__(self, vllm_config: VllmConfig, device: torch.device):
         super().__init__(vllm_config, device)
 
-        # Anchor-first: N query tokens per request (anchor + N-1 noise), not 1+N.
-        self.num_query_per_req = self.num_speculative_steps
+        # Anchor-first (N slots) vs. speculators "1+N" fill-in (block_size==N+1).
+        block_size = getattr(self.draft_model_config.hf_config, "block_size", None)
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: This seems fragile. Won't this change the mechanism if I use a speculators checkpoint with a different num_speculative_steps?

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 5]
```diff
@@ -38,8 +40,15 @@ class DSparkSpeculator(DFlashSpeculator):
...
+        self.anchor_as_first = (
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: What's the point of renaming and duplicating `sample_from_anchor`? I prefer `sample_from_anchor` as a name and I'm not sure why we need two variables

File: vllm/model_executor/models/qwen3_dspark.py (hunk 2 above)
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dspark.py:137]: Does the deepseek_v4/nvidia/dspark.py need to implement these functions as well?

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 6]
```diff
@@ -94,9 +125,20 @@ def _sample_sequential(self, num_reqs: int, head_hidden: torch.Tensor) -> None:
...
             if self.draft_logits is not None:
+                # Probabilistic: sample in target vocab so the rejection sampler
+                # can index the stored logits by target id (the sampled id is
+                # then already a target id). A reduced draft vocab is scattered
+                # into target columns (-inf elsewhere); a full-vocab draft
+                # already lives in target space.
+                if self._d2t_scatter_index is not None:
+                    assert self._draft_scatter_buf is not None
+                    buf = self._draft_scatter_buf[:num_reqs]
+                    buf.fill_(float("-inf"))
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: naive Q, is -inf numerically stable here?

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 7]
```diff
@@ -94,9 +125,20 @@ def _sample_sequential(self, num_reqs: int, head_hidden: torch.Tensor) -> None:
...
+                    buf.index_copy_(1, self._d2t_scatter_index, logits_i.to(buf.dtype))
+                    logits_i = buf
                 # sample_pos is the predicted token's position Q; the target
                 # verifies it with the predecessor's Gumbel key (Q-1). Pass Q-1.
-                draft_i = gumbel_sample(
+                target_i = gumbel_sample(
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: nit: `target_i` is a confusing name. maybe `draft_vocab_i` -> `draft_sampled_i` or something

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 8]
```diff
@@ -94,9 +125,20 @@ def _sample_sequential(self, num_reqs: int, head_hidden: torch.Tensor) -> None:
...
+                # Probabilistic: sample in target vocab so the rejection sampler
+                # can index the stored logits by target id (the sampled id is
+                # then already a target id). A reduced draft vocab is scattered
+                # into target columns (-inf elsewhere); a full-vocab draft
+                # already lives in target space.
+                if self._d2t_scatter_index is not None:
+                    assert self._draft_scatter_buf is not None
+                    buf = self._draft_scatter_buf[:num_reqs]
+                    buf.fill_(float("-inf"))
+                    buf.index_copy_(1, self._d2t_scatter_index, logits_i.to(buf.dtype))
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: Is there some native torch op that will fuse the fill and scatter, since we're only ever assigning a 1-to-1 mapping? maybe a `torch.where` assignment or something?

File: vllm/model_executor/models/qwen3_dspark.py (hunk 2 above)
Comment (mgoin) [reply vllm/model_executor/models/qwen3_dspark.py:137]: Very good catch

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py (hunk 8 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: Since the scatter index is constant, we can just fill once during the init, that way we only scatter at runtime

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py (hunk 4 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: Another good call, I'll just add a special flag in the speculators config translation

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py (hunk 6 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: Yes it should be fine

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 9]
```diff
@@ -94,9 +122,16 @@ def _sample_sequential(self, num_reqs: int, head_hidden: torch.Tensor) -> None:
             bias = self.model.markov_bias(markov_embed)
             logits_i = base_logits[:, i] + bias
             if self.draft_logits is not None:
+                # Probabilistic: sample in target vocab (a reduced draft vocab is
+                # scattered into its target columns; full vocab is already there).
+                if self._d2t_scatter_index is not None:
+                    assert self._draft_scatter_buf is not None
+                    buf = self._draft_scatter_buf[:num_reqs]
+                    buf.index_copy_(1, self._d2t_scatter_index, logits_i.to(buf.dtype))
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py:130]: what happened to the `-inf` fill? won't the other slots be junk if you don't initialize them?

Comment (benchislett) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py:130]: index map is constant. they never get overridden so it's okay to leave them untouched here

**Review by benchislett — APPROVED** [review]
LGTM

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py (hunk 9 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py:130]: we now -inf once in _draft_scatter_buf during load. since we are always scattering to the same places, i realized we don’t need to fill
